crates/network/src/transaction/builder.rs

```diff
@@ -0,0 +1,200 @@
+use super::signer::NetworkSigner;
+use crate::Network;
+use alloy_primitives::{Address, Bytes, ChainId, TxKind, U256, U64};
+use async_trait::async_trait;
+
+/// Error type for transaction builders.
+#[derive(Debug, thiserror::Error)]
+pub enum TransactionBuilderError {
+    /// A required key is missing.
+    #[error("A required key is missing: {0}")]
+    MissingKey(&'static str),
+
+    /// Signer cannot produce signature type required for transaction.
+    #[error("Signer cannot produce signature type required for transaction")]
+    UnsupportedSignatureType,
+
+    /// Signer error.
+    #[error(transparent)]
+    Signer(#[from] alloy_signer::Error),
+
+    /// A custom error.
+    #[error("{0}")]
+    Custom(#[source] Box<dyn std::error::Error + Send + Sync + 'static>),
+}
+
+impl TransactionBuilderError {
+    /// Instantiate a custom error.
+    pub fn custom<E>(e: E) -> Self
+    where
+        E: std::error::Error + Send + Sync + 'static,
+    {
+        Self::Custom(Box::new(e))
+    }
+}
+
+/// [`TransactionBuilder`] result type.
+pub type BuilderResult<T, E = TransactionBuilderError> = std::result::Result<T, E>;
+
+/// A Transaction builder for a network.
+///
+/// Transaction builders are primarily used to construct typed transactions that can be signed with
+/// [`TransactionBuilder::build`], or unsigned typed transactions with
+/// [`TransactionBuilder::build_unsigned`].
+///
+/// Transaction builders should be able to construct all available transaction types on a given
+/// network.
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
```

Comment:
we need it, otherwise we have no way to set input/value etc. in e.g. `CallBuilder`, since all of that is just `N::TransactionRequest`

Reply:
i misread, i thought u said we don't need the trait (missed async) lmao

we need async trait for the build fn that signs the tx since the signers are async, are u suggesting we just do sync signing?

Reply:
tried rm'ing, we still need it cus of some bounds in `Provider` (i assume added by async trait)

Reply:
you have to remove it in impls too

Reply:
let's do in followup

Reply:
https://github.com/alloy-rs/alloy/issues/268